File: kata_09.rb
RULE = { 
    "A" => { "unit_price" => 50, "special_price" => "3 for 130" },
    "B" => { "unit_price" => 30, "special_price" => "2 for 45" },
    "C" => { "unit_price" => 20, "special_price" => nil },
    "D" => { "unit_price" => 15, "special_price" => nil }
}

class CheckOut

    def initialize(rules)
        @rules = rules
        @list = Hash.new(0)
    end

    def scan(item)
        @list[item] += 1
    end

    def total
        total = 0
        @list.each do |key, quantity|
            if !@rules[key]['special_price'].nil?
                special_quantity = @rules[key]['special_price'].split(' ')[0].to_i
                special_value = @rules[key]['special_price'].split(' ')[2].to_i
                quotient = quantity / special_quantity
                remainder = quantity % special_quantity
                total += ((quotient * special_value) + (remainder * @rules[key]['unit_price']))
            else
                total += quantity * @rules[key]['unit_price']
            end
        end
        return total
    end

end

require "test/unit"

class TestPrice < Test::Unit::TestCase
    
    def price(goods)
        co = CheckOut.new(RULE)
        goods.split(//).each { |item| co.scan(item) }
        co.total
    end

    def test_totals
        assert_equal(  0, price(""))
        assert_equal( 50, price("A"))
        assert_equal( 80, price("AB"))
        assert_equal(115, price("CDBA"))

        assert_equal(100, price("AA"))
        assert_equal(130, price("AAA"))
        assert_equal(180, price("AAAA"))
        assert_equal(230, price("AAAAA"))
        assert_equal(260, price("AAAAAA"))

        assert_equal(160, price("AAAB"))
        assert_equal(175, price("AAABB"))
        assert_equal(190, price("AAABBD"))
        assert_equal(190, price("DABABA"))
    end

    def test_incremental
        co = CheckOut.new(RULE)
        assert_equal(0, co.total)
        co.scan("A"); assert_equal( 50, co.total)
        co.scan("B"); assert_equal( 80, co.total)
        co.scan("A"); assert_equal(130, co.total)
        co.scan("A"); assert_equal(160, co.total)
        co.scan("B"); assert_equal(175, co.total)
    end
end

File: kata09.py
PRICING_RULES = {
    'A': [50, '3 for 130'],
    'B': [30, '2 for 45'],
    'C': [20, ''],
    'D': [15, ''],
}


class Checkout:
    def __init__(self, pricing_rules):
        self.total = 0
        self.pricing_rules = pricing_rules
        self.stock = {}
        for key in self.pricing_rules.keys():
            self.stock[key] = 0

    def update_total(self):
        total = 0
        for key, rule in self.pricing_rules.items():
            quantity = self.stock[key]
            unit_price, special_offer = rule
            if special_offer == '':
                total += (unit_price * quantity)
            else:
                special_quantity, special_price = [int(value) for value in special_offer.split(" for ")]
                quotient = quantity // special_quantity
                remainder = quantity % special_quantity
                total += ((special_price * quotient) + (unit_price * remainder))
        self.total = total

    def scan(self, item):
        self.stock[item] += 1
        self.update_total()

    @staticmethod
    def new(pricing_rules):
        co = Checkout(pricing_rules)
        return co


class TestPrice:
    @staticmethod
    def price(goods):
        co = Checkout.new(PRICING_RULES)
        for item in goods: 
            co.scan(item)
        total = co.total
        return total

    @staticmethod
    def test_totals():
        assert TestPrice.price("") == 0, "Alguma coisa deu errado!"
        assert TestPrice.price("A") == 50, "Alguma coisa deu errado!"
        assert TestPrice.price("AB") == 80, "Alguma coisa deu errado!"
        assert TestPrice.price("CDBA") == 115, "Alguma coisa deu errado!"

        assert TestPrice.price("AA") == 100, "Alguma coisa deu errado!"
        assert TestPrice.price("AAA") == 130, "Alguma coisa deu errado!"
        assert TestPrice.price("AAAA") == 180, "Alguma coisa deu errado!"
        assert TestPrice.price("AAAAA") == 230, "Alguma coisa deu errado!"
        assert TestPrice.price("AAAAAA") == 260, "Alguma coisa deu errado!"

        assert TestPrice.price("AAAB") == 160, "Alguma coisa deu errado!"
        assert TestPrice.price("AAABB") == 175, "Alguma coisa deu errado!"
        assert TestPrice.price("AAABBD") == 190, "Alguma coisa deu errado!"
        assert TestPrice.price("DABABA") == 190, "Alguma coisa deu errado!"

    @staticmethod
    def test_incremental():
        co = Checkout.new(PRICING_RULES)
        assert co.total == 0, "Alguma coisa deu errado!"
        co.scan("A")
        assert co.total == 50, "Alguma coisa deu errado!"
        co.scan("B")
        assert co.total == 80, "Alguma coisa deu errado!"
        co.scan("A")
        assert co.total == 130, "Alguma coisa deu errado!"
        co.scan("A")
        assert co.total == 160, "Alguma coisa deu errado!"
        co.scan("B")
        assert co.total == 175, "Alguma coisa deu errado!"


if __name__ == "__main__":
    TestPrice.test_totals()
    TestPrice.test_incremental()
